### DISCUSSION TITLE
Load relationship/s to an already loaded object

### DISCUSSION THREAD

**Discussion by omni-dispatch-dev** [sqlalchemy/sqlalchemy#11901] [Usage Questions] [answered]
Hey, could you explain how can I load relationship with populate_existing for already loaded objects? 

There is simmilar question https://github.com/sqlalchemy/sqlalchemy/discussions/8719 but there is no code example and it's unclear for me.

I found a solution for refresh, where I can load relationship for one object:

```
room = (await session.execute(select(School))).scalars().one()
# some staff here
classes = await session.refresh(School, attribute_names=["classes"])
But what do I do when I need to load relations for many/nested objects (probably with populate_existing)?

schools = (await session.execute(
    select(School).options(
        joinedload(School.clesses)
    )
)).scalars().all()
# some staff here
# and now i need to load people for each class in each school, how do i do that?  
```

I found a way to load a relation for one school:

`await session.get(School, [1], options=(joinedload(School.classes).options(joinedload(Class.people)),), populate_existing=True)`

But it is for only one object in session, is there any way to do it for list of objects?

Thanks

**Comment by zzzeek** [comment]
this gets into the realm of why not just call `session.expire()` ?   If you have a list of objects for which you've done some kind of out-of-band UPDATE statement, or what else is changing on these ?    it gets complicated to surgically expire a sub-graph of objects.   

if you wanted to port your `session.get` version to a list of objects then you would use  `select().where(School.id.in([list of ids])).options(...)` .    but this seems like the top level assumptions might want to be revised if possible.

**Reply by omni-dispatch-dev** [reply]
> this gets into the realm of why not just call `session.expire()` ? 
Because there were no changes to those objects.
Let me explain it in provided example:

```
schools = (await session.execute(
    select(School).options(
        joinedload(School.clesses)
    ) # I have a list of schools with related clesses
    # Here are some checks and 3rd party calls (request to another service etc)  without any changes to ORM objects
    # and here is condition when I might need to load people to classes. 
    # I don't want to load people in first query because I might not even need them. 
```
To be honest don't  understand what you mean by 
`select().where(School.id.in([list of ids])).options(...)`
because it's just doesn't work. 
But `select(School).where(School.id.in([list of ids])).options(joinedload(School.clesses)...)` issue a brand new query, which I want to avoid (if possible).

**Reply by zzzeek** [reply]
how do you want to joinedload all the related collections without emitting a SELECT for the top level table?       Think about what SQL you want to emit in order to load what you need.

**Reply by omni-dispatch-dev** [reply]
But I already have ids of needed classes: 
For example I have two Schools with ids 1 and 2, and each school has one class with id 3 and 4. 
So I want to emit `SELECT * FROM people  where people.class_id IN [3, 4]` and to link objects (people) in result to existing classes by 
expression `people.class_id == class.id `

**Reply by zzzeek** [reply, marked as answer]
if you want to avoid the top SELECT and do a "selectinload" for the related items only you want to use [this recipe](https://github.com/sqlalchemy/sqlalchemy/wiki/DisjointEagerLoading).

Basically emit the SELECT query you want, group the results by parent id, then use [set_committed_value](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.attributes.set_committed_value) to assign the new collections to the parent objects

**Reply by omni-dispatch-dev** [reply]
Yeah, thank you, that's exactly what I was looking for. 

Thank you again
